src/osd/OSD.cc

```diff
@@ -7635,9 +7635,14 @@ void OSD::_committed_osd_maps(epoch_t first, epoch_t last, MOSDMap *m)
     else
       start_boot();
   }
-  else if (do_restart)
+  else if (do_restart) {
+    vector<PGRef> pgs;
+    _get_pgs(&pgs);
+    for (auto pg : pgs) {
+      pg->force_restart_peering();
```

Comment:
would it be sufficient o call handle_initialize() here? That will transition us to Reset...

Reply:
I tried handle_initialize().
Unfortunately, the osd crashed here since no transition like this:
>    Started/ReplicaActive/RepNotRecovering ---('Initialize' event)---> Reset

What about the patch below instead? This modifies only one portion.
If you prefer this, I'll remake the PR.

```
--- a/src/osd/OSD.cc
+++ b/src/osd/OSD.cc
@@ -7665,9 +7665,18 @@ void OSD::_committed_osd_maps(epoch_t first, epoch_t last, MOSDMap *m)
     else
       start_boot();
   }
-  else if (do_restart)
-    start_boot();
+  else if (do_restart) {
+    vector<PGRef> pgs;
+    _get_pgs(&pgs);
+    for (auto pg : pgs) {
+      vector<int> acting;
+      // force restart a peering
+      osdmap->pg_to_up_acting_osds(pg->pg_id.pgid, nullptr, nullptr, &acting, nullptr);
+      service.queue_want_pg_temp(pg->pg_id.pgid, twiddle(acting), true);
+    }
 
+    start_boot();
+  }
 }
 
 void OSD::check_osdmap_features()
```
I think the previous patch is straightforward, though.

Reply:
I tested the twiddle acting patch. It might not work well.
So far, I don't have any other idea to reset a PG.